### PR TITLE
Refine ruff lint configuration and update tooling to Python 3.12

### PR DESCRIPTION
**Pull request by mohavro**
### Motivation
- Align project tooling with Python 3.12 and resolve reviewer feedback about incorrectly-placed Ruff settings.
- Ensure `isort`/first-party module detection is configured for Ruff formatting and unify dev tooling in `pyproject.toml` and `requirements-dev.txt`.

### Description
- Updated project runtime to `requires-python = ">=3.12"` and removed older 3.10/3.11 classifiers in `pyproject.toml`.
- Moved Ruff selectors into the supported `[tool.ruff.lint]` table and added `[tool.ruff.lint.isort]` with `known-first-party = ["src"]` to align Ruff with `isort` behavior.
- Added Ruff, `autopep8`, and `radarlint` to the `dev` extras in `pyproject.toml` and to `requirements-dev.txt`.
- Adjusted formatter/typing tool targets to Python 3.12 by changing `tool.black.target-version`, `tool.mypy.python_version`, and `tool.pylint.py-version` to `3.12`.

### Testing
- No automated tests or linters were executed as part of this change; CI should run the normal test and linting pipelines after merging.

[Codex Task](https://chatgpt.com/codex/tasks/task_e_6988d0a84b78832498a2b013b6b81b97)

### **Auto-created Ticket**

[#800](https://github.com/DashFin-FarDb/financial-asset-relationship-db/issues/800)

### **PR Type**
Enhancement

### **Description**
- Upgrade minimum Python requirement to 3.12

- Add Ruff linter with comprehensive configuration

- Update formatter/typing tools to target Python 3.12

- Add autopep8 and radarlint to dev dependencies

### Diagram Walkthrough

```mermaid
flowchart LR
  A["Python 3.10/3.11"] -->|"Upgrade to"| B["Python 3.12"]
  B --> C["Update tool targets"]
  C --> D["Black, MyPy, Pylint"]
  E["Add dev tools"] -->|"Ruff, autopep8"| F["Enhanced linting"]
  G["Ruff config"] -->|"isort integration"| H["known-first-party src"]
```

<details><summary><h3>File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Configuration changes</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>pyproject.toml</strong><dd><code>Python 3.12 upgrade and Ruff linter setup</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pyproject.toml

<ul><li>Updated <code>requires-python</code> from <code>>=3.10</code> to <code>>=3.12</code><br> <li> Removed Python 3.10 and 3.11 classifiers, kept only 3.12<br> <li> Added <code>ruff>=0.4.0</code>, <code>autopep8>=2.0.0</code>, and <code>radarlint>=0.1.0</code> to dev extras<br> <li> Updated <code>tool.black.target-version</code> to <code>['py312']</code><br> <li> Added comprehensive <code>[tool.ruff]</code> configuration with lint, isort, and <br>format sections<br> <li> Updated <code>tool.mypy.python_version</code> and <code>tool.pylint.py-version</code> to <code>3.12</code><br> <li> Added <code>[tool.flake8]</code> and <code>[tool.autopep8]</code> configuration sections</ul>

</details>

  </td>
  <td><a href="https://github.com/DashFin-FarDb/financial-asset-relationship-db/pull/798/files#diff-50c86b7e
...(truncated)

___

## **CodeAnt-AI Description**
Expose formula.expression while renaming internal field and make correlation averaging robust

### What Changed
- Formula objects now store the textual expression as "formula" but expose the old "expression" attribute via a read-only property so existing consumers keep working without changes
- Average correlation calculation now accepts mixed input formats (numbers and nested dicts), extracts numeric values safely, ignores exact-1.0 entries, and falls back to 0.5 when no valid values are present
- Added a large suite of unit and integration tests (app, MCP server, asset graph, Makefile targets, visualizations, repository and data fetcher) to cover UI flows, server concurrency, graph edge cases, data fetching edge conditions, and Makefile targets

### Impact
 `✅ No breaks for users relying on formula.expression`
 `✅ Fewer crashes on malformed or nested correlation data`
 `✅ Higher confidence from expanded unit and integration tests`
<details>
<summary><strong>💡 Usage Guide</strong></summary>

### Checking Your Pull Request
Every time you make a pull request, our system automatically looks through it. We check for security issues, mistakes in how you're setting up your infrastructure, and common code problems. We do this to make sure your changes are solid and won't cause any trouble later.

### Talking to CodeAnt AI
Got a question or need a hand with something in your pull request? You can easily get in touch with CodeAnt AI right here. Just type the following in a comment on your pull request, and replace "Your question here" with whatever you want to ask:
<pre>
<code>@codeant-ai ask: Your question here</code>
</pre>
This lets you have a chat with CodeAnt AI about your pull request, making it easier to understand and improve your code.

#### Example
<pre>
<code>@codeant-ai ask: Can you suggest a safer alternative to storing this secret?</code>
</pre>

### Preserve Org Learnings with CodeAnt
You can record team preferences so CodeAnt AI applies them in future reviews. Reply directly to the specific CodeAnt AI suggestion (in the same thread) and replace "Your feedback here" with your input:
<pre>
<code>@codeant-ai: Your feedback here</code>
</pre>
This helps CodeAnt AI learn and adapt to your team's coding style and standards.

#### Example
<pre>
<code>@codeant-ai: Do not flag unused imports.</code>
</pre>

### Retrigger review
Ask CodeAnt AI to review the PR again, by typing:
<pre>
<code>@codeant-ai: review</code>
</pre>

### Check Your Repository Health
To analyze the health of your code repository, visit our dashboard at [https://app.codeant.ai](https://app.codeant.ai). This tool helps you identify potential issues and areas for improvement in your codebase, ensuring your repository maintains high standards of code health.

</details>

### CODE REVIEW

**Comment by mohavro** [comment]
### :white_check_mark: **Snyk checks have passed. No issues have been found so far.**

| Status | Scanner |  Critical |  High |  Medium |  Low | Total (0) |
|---|---|---|---|---|---|---|
| :white_check_mark: | Open Source Security| 0 | 0 | 0 | 0 | [ 0 issues](https://app.snyk.io/org/dashfin.fardb/pr-checks/c6557798-319c-4bcf-8392-3df3702f1fe7?refs=5e819a35de41b01adb4c53d7af525423e4c8fed2&source=prChecksComment) |

:computer: Catch issues earlier using the plugins for [VS Code](https://marketplace.visualstudio.com/items?itemName=snyk-security.snyk-vulnerability-scanner), [JetBrains IDEs](https://plugins.jetbrains.com/plugin/10972-snyk-security), [Visual Studio](https://marketplace.visualstudio.com/items?itemName=snyk-security.snyk-vulnerability-scanner-vs-2022), and [Eclipse](https://marketplace.eclipse.org/content/snyk-security#external-install-button).

File: requirements-dev.txt [hunk 1]
```diff
@@ -7,6 +7,9 @@ pylint>=2.17.0
 mypy>=1.0.0
 black>=23.0.0
 isort>=5.12.0
+ruff>=0.4.0
+autopep8>=2.0.0
+radarlint>=0.1.0
```
[the remark this answers is not in the document]
Comment (mohavro) [reply requirements-dev.txt]: @codex review and resolve this comment

File: .github/workflows/autofix.yml [hunk 2]
```diff
@@ -13,14 +13,20 @@ jobs:
     steps:
       - uses: actions/checkout@v6
 
-      # Use uv to ensure we have the same ruff version in CI and locally.
-      - uses: astral-sh/setup-uv@681c641aba71e4a1c380be3ab5e12ad51f415867
+      - name: Set up Python
+        uses: actions/setup-python@<PINNED_SHA_FOR_V5>
```
[the remark this answers is not in the document]
Comment (mohavro) [reply .github/workflows/autofix.yml]:
```suggestion
        uses: actions/setup-python@<PINNED_SHA_FOR_V5>
        with:
          python-version: '3.12'
```

File: pyproject.toml [hunk 3]
```diff
@@ -106,8 +107,33 @@ line_length = 120
...
+line-length = 120
+target-version = "py312"
+fix = true
+unsafe-fixes = true
+
+[tool.ruff.lint]
+extend-select = ["I", "UP", "B", "C4", "SIM"]
+
+[tool.ruff.lint.isort]
+known-first-party = ["src"]
```
[the remark this answers is not in the document]
Comment (mohavro) [reply pyproject.toml]: Can you suggest code to fix this issue?

File: pyproject.toml [hunk 4]
```diff
@@ -100,14 +98,36 @@ extend-exclude = '''
...
-skip_gitignore = true
-known_first_party = ["src"]
+
+[tool.ruff]
+fix = true
+unsafe-fixes = false
+line-length = 120
+target-version = "py312"
+fix = true
+unsafe-fixes = false
```
[the remark this answers is not in the document]
Comment (mohavro) [reply pyproject.toml]:
```suggestion
[tool.ruff]
line-length = 120
target-version = "py312"
fix = true
unsafe-fixes = false
```

File: tests/integration/test_makefile.py [hunk 5]
```diff
@@ -0,0 +1,447 @@
...
+        assert result.returncode == 0, "lint target not found"
+
+    @staticmethod
+    def test_lint_target_uses_linters():
+        """Test that 'lint' target uses linting tools."""
+        result = subprocess.run(["make", "-n", "lint"], capture_output=True, text=True, check=False)
+
+        output = result.stdout + result.stderr
+        found_linters = [linter for linter in linters if linter in output]
+        assert len(found_linters) > 0, "lint target doesn't use any known linters"
```
[the remark this answers is not in the document]
Comment (mohavro) [reply tests/integration/test_makefile.py]:
```suggestion
    @staticmethod
    def test_lint_target_uses_linters():
        """Test that 'lint' target uses linting tools."""
        result = subprocess.run(["make", "-n", "lint"], capture_output=True, text=True, check=False)

        output = result.stdout + result.stderr
        linters = ["ruff", "flake8", "pylint", "mypy", "bandit"]
        found_linters = [linter for linter in linters if linter in output]
        assert len(found_linters) > 0, "lint target doesn't use any known linters"
```

File: src/analysis/formulaic_analysis.py [hunk 6]
```diff
@@ -26,14 +26,18 @@ class Formula:
...
     latex: str
     description: str
     variables: Dict[str, str]  # variable_name -> description
     example_calculation: str
     category: str
     r_squared: float = 0.0  # Correlation strength if applicable
 
+    @property
+    def expression(self) -> str:
+        return self.formula
```
[the remark this answers is not in the document]
Comment (mohavro) [reply src/analysis/formulaic_analysis.py]:
```suggestion
        return self.formula
        
    @expression.setter
    def expression(self, value: str) -> None:
        self.formula = value
```

File: tests/integration/test_makefile.py [hunk 7]
```diff
@@ -0,0 +1,447 @@
...
+
+    @staticmethod
+    def test_lint_target_uses_linters():
+        """Test that 'lint' target uses linting tools."""
+        result = subprocess.run(["make", "-n", "lint"], capture_output=True, text=True, check=False)
+
+        output = result.stdout + result.stderr
+        linters = ["ruff", "flake8", "pylint", "mypy", "bandit"]
+        found_linters = [linter for linter in linters if linter in output]
+        assert len(found_linters) > 0, "lint target doesn't use any known linters"
```
[the remark this answers is not in the document]
Comment (mohavro) [reply tests/integration/test_makefile.py]:
```suggestion
        linters = ["ruff"]
        output_lc = output.lower()
        found_linters = [linter for linter in linters if linter in output_lc]
        assert result.returncode == 0 and (found_linters or output.strip()), "lint target doesn't appear to invoke a linter"
```

File: tests/integration/test_requirements_dev.py [hunk 8]
```diff
@@ -606,14 +609,10 @@ def test_has_type_checker(package_names: List[str]):
...
-        """Test that an import sorter is present."""
-        import_sorters = ["isort", "reorder-python-imports"]
-        # This is optional but good to have
-        if any(sorter in package_names for sorter in import_sorters):
-            assert True
-        else:
-            # Not required but log it
-            pass
+        """Test that an import sorter is present (optional but recommended)."""
+    import_sorters = ["ruff", "isort", "reorder-python-imports"]
```
[the remark this answers is not in the document]
Comment (mohavro) [reply tests/integration/test_requirements_dev.py]:
```suggestion
def test_has_type_checker(package_names: List[str]):
    """Test that an import sorter is present (optional but recommended)."""
    import_sorters = ["ruff", "isort", "reorder-python-imports"]
    assert any(sorter in package_names for sorter in import_sorters)
    import_sorters = ["ruff", "isort", "reorder-python-imports"]
    if not any(sorter in package_names for sorter in import_sorters):
        pytest.skip("No import-sorting tool found (optional)")
```

**Review by mohavro — COMMENTED** [review]
approved

**Comment by mohavro** [comment]
## Closing — branch has syntax errors in source files and is stale against main

After full review, this PR cannot be merged. Three of its changed files have Python syntax errors confirmed by `py_compile`:

| File | Error |
|------|-------|
| `api/main.py` line 38 | `SyntaxError: expected 'except' or 'finally' block` — a `try:` block was left without an exception handler after an automated fix |
| `src/reports/schema_report.py` line 124 | `SyntaxError: 'return' outside function` — de-indentation pushed the return statement to module level |
| `tests/unit/test_mcp_server.py` line 129 | `IndentationError: unexpected indent` — a class docstring string is floating at the wrong indentation level |

Additionally:
- The branch is `CONFLICTING`/`DIRTY` against `main` (130 commits ahead, conflicts across 40 files)
- `pyproject.toml` has `target-version = "py312"` duplicated twice in the `[tool.ruff]` section
- `radarlint>=0.1.0` is added as a dev dependency — an unrecognised package not warranted by the stated goals

### Salvageable config changes

The legitimate improvements (Python 3.12 targets in black/mypy/pylint, isort→ruff migration in pre-commit and requirements-dev.txt) will be applied cleanly to `main` in a separate commit to preserve the value without the broken code.
